File: src/store.ts
/**
 * Estado que sobrevive al cierre: historial de versiones y sesión.
 *
 * Vivía en `localStorage` y ahí no cabe. Treinta copias de un documento de
 * diez mil líneas son catorce megas contra una cuota de cinco, y al llenarse
 * `history.ts` se tragaba el error: la aplicación seguía ofreciendo «recuperar
 * versión» sin tener ninguna. La sesión compartía esa cuota, así que un
 * documento grande podía impedir guardar los borradores sin guardar.
 *
 * En el escritorio va a disco, un archivo por cosa, con escritura atómica. En
 * el navegador —`npm run dev`— sigue en `localStorage`, que es lo único que
 * hay allí; y allí los documentos de prueba son pequeños.
 */

import { invoke } from "@tauri-apps/api/core";
import { isTauri } from "./files.ts";

export type StoreName = "history" | "session";

const CLAVE_ANTIGUA: Record<StoreName, string> = {
  history: "unfold:history",
  session: "unfold:session",
};

export interface Store {
  read(name: StoreName): Promise<string | null>;
  write(name: StoreName, contents: string): Promise<void>;
  clear(name: StoreName): Promise<void>;
}

const enDisco: Store = {
  read: (name) => invoke<string | null>("store_read", { name }),
  write: async (name, contents) => {
    await invoke("store_write", { name, contents });
  },
  clear: async (name) => {
    await invoke("store_clear", { name });
  },
};

const enNavegador: Store = {
  read: async (name) => localStorage.getItem(CLAVE_ANTIGUA[name]),
  write: async (name, contents) => localStorage.setItem(CLAVE_ANTIGUA[name], contents),
  clear: async (name) => localStorage.removeItem(CLAVE_ANTIGUA[name]),
};

export const store: Store = isTauri ? enDisco : enNavegador;

/**
 * Sube a disco lo que hubiera quedado en `localStorage`.
 *
 * Se hace una vez, al arrancar, y sólo si en disco no hay nada: nadie debe
 * perder su historial por actualizar. La clave vieja se borra después, que es
 * además lo que libera la cuota que estaba causando el problema.
 */
export async function migrarDesdeLocalStorage(destino: Store = store): Promise<void> {
  if (!isTauri) return;
  for (const name of ["history", "session"] as StoreName[]) {
    const viejo = localStorage.getItem(CLAVE_ANTIGUA[name]);
    if (viejo === null) continue;
    try {
      if ((await destino.read(name)) === null) await destino.write(name, viejo);
      localStorage.removeItem(CLAVE_ANTIGUA[name]);
    } catch (error) {
      // Si la subida falla se deja donde estaba: mejor un historial en la
      // cuota vieja que ningún historial.
      console.error(`No se pudo migrar ${name} a disco`, error);
    }
  }
}

File: src-tauri/src/store.rs
//! Documentos de estado guardados en disco.
//!
//! El historial de versiones y la sesión vivían en `localStorage`, y ahí no
//! caben: treinta copias completas de un documento de diez mil líneas son
//! catorce megas contra una cuota de cinco. Al llenarse, escribir falla —y
//! `history.ts` se lo tragaba en silencio, de modo que la aplicación seguía
//! ofreciendo «recuperar versión» sin tener ninguna guardada.
//!
//! Peor todavía: la sesión compartía esa misma cuota con el historial, así que
//! un documento grande podía impedir guardar los borradores sin guardar.
//!
//! En disco no hay cuota, y cada cosa va en su archivo. Se escribe igual que
//! los catálogos: a un temporal y luego renombrar, que en Windows es atómico y
//! sobrescribe, para no dejar nunca un archivo a medias.

use serde::Serialize;
use std::path::PathBuf;
use tauri::{AppHandle, Manager};

/// Qué se puede guardar. Es una lista cerrada a propósito: el nombre llega
/// desde el frontend y sin ella sería una ruta que alguien podría torcer.
const CONOCIDOS: [&str; 2] = ["history", "session"];

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StoreWrite {
    /// Bytes que ocupa ya en disco, para poder decirlo cuando importe.
    pub bytes: u64,
}

fn ruta(app: &AppHandle, name: &str) -> Result<PathBuf, String> {
    if !CONOCIDOS.contains(&name) {
        return Err(format!("«{name}» no es un documento de estado conocido"));
    }
    Ok(app
        .path()
        .app_local_data_dir()
        .map_err(|error| format!("No se pudo localizar la carpeta de Unfold: {error}"))?
        .join(format!("{name}.json")))
}

#[tauri::command]
pub async fn store_read(app: AppHandle, name: String) -> Result<Option<String>, String> {
    let path = ruta(&app, &name)?;
    match std::fs::read_to_string(&path) {
        Ok(contenido) => Ok(Some(contenido)),
        // Que no exista es el primer arranque, no un error.
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => Ok(None),
        Err(error) => Err(format!("No se pudo leer {name}: {error}")),
    }
}

#[tauri::command]
pub async fn store_write(
    app: AppHandle,
    name: String,
    contents: String,
) -> Result<StoreWrite, String> {
    let path = ruta(&app, &name)?;
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent)
            .map_err(|error| format!("No se pudo crear la carpeta de Unfold: {error}"))?;
    }

    let temporal = path.with_extension("json.tmp");
    std::fs::write(&temporal, &contents)
        .map_err(|error| format!("No se pudo escribir {name}: {error}"))?;
    std::fs::rename(&temporal, &path)
        .map_err(|error| format!("No se pudo guardar {name}: {error}"))?;

    Ok(StoreWrite {
        bytes: contents.len() as u64,
    })
}

#[tauri::command]
pub async fn store_clear(app: AppHandle, name: String) -> Result<(), String> {
    let path = ruta(&app, &name)?;
    match std::fs::remove_file(&path) {
        Ok(()) => Ok(()),
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => Ok(()),
        Err(error) => Err(format!("No se pudo borrar {name}: {error}")),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// El nombre viene del frontend. Sin lista cerrada sería una ruta, y
    /// `../../algo` escribiría fuera de la carpeta de la aplicación.
    #[test]
    fn only_known_documents_are_addressable() {
        for malo in ["../secreto", "..", "history/../..", "cualquiera"] {
            assert!(
                !CONOCIDOS.contains(&malo),
                "«{malo}» no debería estar en la lista"
            );
        }
        assert!(CONOCIDOS.contains(&"history"));
        assert!(CONOCIDOS.contains(&"session"));
    }
}
